data/data-pipeline/data_pipeline/etl/runner.py

```diff
@@ -32,38 +33,89 @@ def get_datasets_to_run(dataset_to_run: str):
 
 
-def etl_runner(dataset_to_run: str = None) -> None:
-    """Runs all etl processes or a specific one
+def run_etl_single_dataset(dataset: str = None) -> str:
+    """Runs a specific dataset
 
     Args:
-        dataset_to_run (str): Run a specific ETL process. If missing, runs all processes (optional)
+        dataset (str): A specific dataset eligible for a specific ETL process.
 
     Returns:
         None
     """
-    dataset_list = get_datasets_to_run(dataset_to_run)
+    etl_module = importlib.import_module(
+        f"data_pipeline.etl.sources.{dataset['module_dir']}.etl"
+    )
+    etl_class = getattr(etl_module, dataset["class_name"])
+    etl_instance = etl_class()
 
-    # Run the ETLs for the dataset_list
-    for dataset in dataset_list:
-        etl_module = importlib.import_module(
-            f"data_pipeline.etl.sources.{dataset['module_dir']}.etl"
-        )
-        etl_class = getattr(etl_module, dataset["class_name"])
-        etl_instance = etl_class()
+    etl_module = importlib.import_module(
+        f"data_pipeline.etl.sources.{dataset['module_dir']}.etl"
+    )
+    etl_class = getattr(etl_module, dataset["class_name"])
+    etl_instance = etl_class()
 
-        # run extract
-        etl_instance.extract()
+    # run extract
+    etl_instance.extract()
 
-        # run transform
-        etl_instance.transform()
+    # run transform
+    etl_instance.transform()
 
-        # run load
-        etl_instance.load()
+    # run load
+    etl_instance.load()
 
-        # cleanup
-        etl_instance.cleanup()
+    # cleanup
+    etl_instance.cleanup()
+
+    message = f"Finished ETL for {dataset}"
+
+    return message
+
+
+def etl_runner(dataset_to_run: str = None) -> None:
+    """Runs all etl processes or a specific one
+
+    Args:
+        dataset_to_run (str): Run a specific ETL process. If missing, runs all processes (optional)
+
+    Returns:
+        None
+    """
+    dataset_list = get_datasets_to_run(dataset_to_run)
+    NUMBER_OF_THREADS = 10
+    futures_list = []
+    results = []
+
+    with concurrent.futures.ThreadPoolExecutor(
+        max_workers=NUMBER_OF_THREADS
+    ) as executor:
+        for dataset in dataset_list:
+            futures = executor.submit(
+                # manually create Future object
+                # previously using map, we were
+                # not managing exceptions so thoughtfully
+                run_etl_single_dataset,
+                dataset,
+            )
+            futures_list.append(futures)
+
+        for future in futures_list:
+            try:
+                # emprically tested timeout
+                result = future.result(timeout=300)
+                results.append(result)
+            # this catches any exception for that given dataset
+            # one could customize this to specify which dataset
+            # but we perform so much logging that this may be
+            # corroborating evidence
+            except Exception:
```

Comment:
Ooh, I really don't think we want to do this -- right @sverchdotgov ? If we catch all the exceptions, then we're swallowing errors and the scoring process will proceed without realizing that its data is somehow messed up. 

Am I misunderstanding?

Reply:
You are understanding this correctly Lucas. In the current implementation, we do not bubble this exception up from the call stack. This would cause downstream effects on the data pipeline. I did this primarily because I was not sure if we stop execution or skip over the dataset and then handle the dataset separately. 

We can accommodate both errors and results callbacks to handle both workflows downstream.  Depending on our decision we can 1)  cancel everything or just handle the successful results. In the latter case, we would have to refactor the processing pipeline to only look for files that were ETL'ed properly.

Reply:
Yeah we definitely need to cancel everything if there's an error. We can look at re-running automatically – https://github.com/usds/justice40-tool/issues/1037 – but that would be a separate issue/pull request. 

I think Shaun already answered your question fairly succinctly above: 

> The only thing that I would ask for is to verify that if there is any kind of failure in the pipeline, that doesn't result in a partial failure or some kind of unknown state in the final output. Either it all fails, or all succeeds (or partially succeeds in some well defined way, if that's what we decide).